cmd/kubeadm/app/phases/controlplane/manifests.go

```diff
@@ -132,20 +132,22 @@ func CreateStaticPodFiles(manifestDir, patchesDir string, cfg *kubeadmapi.Cluste
 // getAPIServerCommand builds the right API server command from the given config object and version
 func getAPIServerCommand(cfg *kubeadmapi.ClusterConfiguration, localAPIEndpoint *kubeadmapi.APIEndpoint) []string {
 	defaultArguments := map[string]string{
-		"advertise-address":               localAPIEndpoint.AdvertiseAddress,
-		"insecure-port":                   "0",
-		"enable-admission-plugins":        "NodeRestriction",
-		"service-cluster-ip-range":        cfg.Networking.ServiceSubnet,
-		"service-account-key-file":        filepath.Join(cfg.CertificatesDir, kubeadmconstants.ServiceAccountPublicKeyName),
-		"client-ca-file":                  filepath.Join(cfg.CertificatesDir, kubeadmconstants.CACertName),
-		"tls-cert-file":                   filepath.Join(cfg.CertificatesDir, kubeadmconstants.APIServerCertName),
-		"tls-private-key-file":            filepath.Join(cfg.CertificatesDir, kubeadmconstants.APIServerKeyName),
-		"kubelet-client-certificate":      filepath.Join(cfg.CertificatesDir, kubeadmconstants.APIServerKubeletClientCertName),
-		"kubelet-client-key":              filepath.Join(cfg.CertificatesDir, kubeadmconstants.APIServerKubeletClientKeyName),
-		"enable-bootstrap-token-auth":     "true",
-		"secure-port":                     fmt.Sprintf("%d", localAPIEndpoint.BindPort),
-		"allow-privileged":                "true",
-		"kubelet-preferred-address-types": "InternalIP,ExternalIP,Hostname",
+		"advertise-address":                localAPIEndpoint.AdvertiseAddress,
+		"insecure-port":                    "0",
+		"enable-admission-plugins":         "NodeRestriction",
+		"service-cluster-ip-range":         cfg.Networking.ServiceSubnet,
+		"service-account-key-file":         filepath.Join(cfg.CertificatesDir, kubeadmconstants.ServiceAccountPublicKeyName),
+		"service-account-signing-key-file": filepath.Join(cfg.CertificatesDir, kubeadmconstants.ServiceAccountPrivateKeyName),
+		"service-account-issuer":           "https://kubernetes.default.svc",
```

Comment:
This value as the default is the last open question I have. Continuing the conversation started in https://github.com/kubernetes/kubernetes/pull/95667/files#r510954650

@liggitt:

> I expected this to be the public address of the API server. I think the test value set in local-up-cluster is not one we'd use in a real server. @mikedanese, do we have good guidance for what to set this to? 
> 
> Because server names can change and people can make mistakes and want to fix them, we also need to make sure it is possible for users to modify the issuer. I think the workflow for that looks like:
> 1. Explicitly set `--api-audiences` to include both the old and new issuer values
> 2. If you have multiple API servers, roll out the `--api-audiences` change
> 3. Modify the `--service-account-issuer` to the new value
> 4. Roll out the updated API manifest
> 
> I don't know how these default values propagate into something user-modifiable. cc @neolit123 for insight there

@neolit123:

> > I expected this to be the public address of the API server
> 
> if this should be the address, then we should set `service-account-issuer` to the same value in:
> `"advertise-address":                localAPIEndpoint.AdvertiseAddress,`
> 
> would it be fine that separate instances of the kube-apiserver in an HA setup to have separate value for `service-account-issuer` (if this is based on the address)?
> 
> > I don't know how these default values propagate into something user-modifiable. cc @neolit123 for insight there
> 
> nowadays, kubeadm allows patching the manifest files before the kubelet starts on control-plane nodes (and creates the static pods), so technically users can patch the manifests with the flag changes they want to. it's also possible to use "phases", i.e. create manifests, edit manifests with text editor, start the pods later.


@zshihang:

> issuer is used as `iss` claim in the token. it would be a URL. https://tools.ietf.org/html/rfc7519#section-4.1.1
> 
> > server names can change and people can make mistakes and want to fix them
> 
> users can specify `api-audiences` if they doing the audiences change.
> 
> also, will also allow users to pass in `service-account-issuer` and the default will still be `https://kubernetes.default.svc`.

Reply:
Digging deeper, I was wrong. We don't currently support changing issuer in a way that allows tokens from the previous issuer to still be considered valid.

I have two concerns with using https://kubernetes.default.svc:

* `svc` is not currently a TLD but could become one in the future, at which point clients that tried to resolve `https://kubernetes.default.svc/.well-known/openid-configuration` would fetch validation data from some random server. This could be mitigated in this PR by defaulting this to `"https://kubernetes.default.svc."+cfg.DNSDomain` which would default to "https://kubernetes.default.svc.cluster.local" which is not resolvable.

* It is not currently possible to change the issuer in a non-disruptive way. This could be mitigated in the future by kube-apiserver by adding support for accepting multiple issuers. It seems likely we'd want to do that.

@neolit123 can you verify that a kubeadm user that wants to override just the `--service-account-issuer` arg is able to do so in a layered way without taking full ownership over the manifests?

Reply:
> svc is not currently a TLD but could become one in the future, at which point clients that tried to resolve https://kubernetes.default.svc/.well-known/openid-configuration would fetch validation data from some random server. This could be mitigated in this PR by defaulting this to "https://kubernetes.default.svc."+cfg.DNSDomain which would default to "https://kubernetes.default.svc.cluster.local" which is not resolvable.

that would be a problem, yes. passing `"https://kubernetes.default.svc.cluster.local"` by default for the time being seems good. the `cluster.local` defaulted DNS field is under [`cfg.Networking.DNSDomain`](https://godoc.org/k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm/v1beta2#Networking)

> @neolit123 can you verify that a kubeadm user that wants to override just the --service-account-issuer arg is able to do so in a layered way without taking full ownership over the manifests?

yes, they could pass this to `kubeadm init` (on cluster creation):
```
apiVersion: ...
kind: ClusterConfiguration
apiServer:
  extraArgs:
    service-account-issuer: "https://some-URL"
```

which would override the default of `https://kubernetes.default.svc.cluster.local`

Reply:
how about just set it as `https://kubernetes.default.svc.cluster.local`? users can overwrite anyway.